# Baseline Model
Now that the data is cleaned, it's time to create a baseline model. 
This will be a simple but effective model that provides a benchmark score.
The simple neural model will need to outperform this baseline to be considered a success.

## Setup and Load of the data

In [2]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report

# Load the CLEANED data
file_path = '../data/reviews_cleaned_sample.csv'
df_clean = pd.read_csv(file_path)

# Drop any rows that might be empty after cleaning
df_clean.dropna(inplace=True)
print(df_clean.head())

                                              review sentiment  \
0  I really liked this Summerslam due to the look...  positive   
1  Not many television shows appeal to quite as m...  positive   
2  The film quickly gets to a major chase scene w...  negative   
3  Jane Austen would definitely approve of this o...  positive   
4  Expectations were somewhat high for me when I ...  negative   

                                      cleaned_review  
0  I really like this summerslam due to the look ...  
1  not many television show appeal to quite as ma...  
2  the film quickly get to a major chase scene wi...  
3  jane austen would definitely approve of this o...  
4  expectation be somewhat high for I when I go t...  


## Split of the Dataset
As a best practice, I will split the data into training set and testing sets.
The model will be trained on 80% of the data and then evaluated on the remaining 20% of unseen data to get a realistic measure of it performance.


In [4]:
# Define  features (X) and  target (y)
x = df_clean['cleaned_review']
y = df_clean['sentiment']

# Split the data
x_train, x_test, y_train , y_test = train_test_split(x, y, test_size=0.2, random_state=42, stratify=y)

print(f"Training set size: {len(x_train)}")
print(f"Testing set size: {len(x_test)}")


Training set size: 800
Testing set size: 200


### Vectorize the text with TF-IDF and N-grams
Will fit the vectorizer only on the training data to learn the vocabulary, and then use it to transform both the training and testing data


In [5]:
# Initialize the TF-IDF Vectorizer with the N-gram strategy
tfidf = TfidfVectorizer(ngram_range=(1, 3)) # Use unigrams, bigrams and trigrams

# Fit and transform the training data
x_train_tfidf= tfidf.fit_transform(x_train)

# Only transform the test data (using the vocabulary learned from the training data)
x_test_tfidf = tfidf.transform(x_test)

print("Text data vectorizer succesfully.")

Text data vectorizer succesfully.


In [6]:
# Initialize a simple and powerful logistic regression model
model = LogisticRegression()

# Train the model on our vectorized training data
print("Training the baseline model...")
model.fit(x_train_tfidf, y_train)
print("Model training complete!")


Training the baseline model...
Model training complete!


## Evaluation of the baseline model

In [9]:
# Make predictions on the unseen test data
y_pred = model.predict(x_test_tfidf)

# The final Report 
print("\n Baseline Model Performance")

print(classification_report(y_test, y_pred))


 Baseline Model Performance
              precision    recall  f1-score   support

    negative       0.67      0.94      0.79       105
    positive       0.89      0.49      0.64        95

    accuracy                           0.73       200
   macro avg       0.78      0.72      0.71       200
weighted avg       0.77      0.73      0.71       200



## Interpretation of results and conclusion

The baseline model achieved a **weighted average F1-score of 0.71**

This performance report reveals a clear 'personality' for the baseline model:
* **Strenght:** The model is excellent at identifying negative reviews, achieving a **Recall of 0.94** for the 'negative' class.
  This means it successfully finds 94% of all truly negative reviews.
* **Weakness:** Its primary weakness is a very low **Recall of 0.49** for the 'positive' class.
  This means it fails to identify more than half of the truly positive review, often misclassifying them as negative.



## Conclusion

The TF-IDF + N-gram model act as a 'pessimist' it aggresivly flags reviews as negative. This results in high recall for the negative class but poor performance on the positive class

The oficial baseline score to beat with the simple neural model is an **F1-Score of 0.71**.